### PR TITLE
avoid socket leak if HTTPSConnection.connect fails

### LINKED ISSUE #2570: socket leaked when ssl_wrap_socket fails
### Subject

socket leaked when ssl_wrap_socket fails 

### Environment

```
>>> import platform
>>> import urllib3
>>> 
>>> print("OS", platform.platform())
OS Linux-5.4.0-100-generic-x86_64-with-glibc2.31
>>> print("Python", platform.python_version())
Python 3.9.10
>>> print("urllib3", urllib3.__version__)
urllib3 1.26.8
```

### Steps to Reproduce

```python
import pytest
import urllib3


def test_requests_leak():
    with pytest.raises(Exception):
        with urllib3.PoolManager(ca_certs=__file__) as http:
            http.request("GET", "https://google.com")
```

### Expected Behavior

no ResourceWarning

### Actual Behavior
```
:

    def unraisable_exception_runtest_hook() -> Generator[None, None, None]:
        with catch_unraisable_exception() as cm:
            yield
            if cm.unraisable:
                if cm.unraisable.err_msg is not None:
                    err_msg = cm.unraisable.err_msg
                else:
                    err_msg = "Exception ignored in"
                msg = f"{err_msg}: {cm.unraisable.object!r}\n\n"
                msg += "".join(
                    traceback.format_exception(
                        cm.unraisable.exc_type,
                        cm.unraisable.exc_value,
                        cm.unraisable.exc_traceback,
                    )
                )
>               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <socket.socket fd=-1, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6>
E               
E               Traceback (most recent call last):
E                 File "/home/graingert/projects/requests-leak/test_requests_leak.py", line 8, in test_requests_leak
E                   http.request("GET", "https://google.com")
E               ResourceWarning: unclosed <socket.socket fd=14, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.99.58', 43618), raddr=('172.217.16.238', 443)>
```

### PR DESCRIPTION
**Pull request by graingert**
Fixes #2570

### CODE REVIEW

File: src/urllib3/connection.py [hunk 1]
```diff
@@ -447,17 +447,16 @@ def connect(self) -> None:
         self._connecting_to_proxy = bool(self.proxy)
 
         sock: Union[socket.socket, "ssl.SSLSocket"]
-        sock = self._new_conn()
+        self.sock = sock = self._new_conn()
```
Comment (graingert) [inline src/urllib3/connection.py:450]: this socket is closed here if `conn.connect()` fails: https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py#L809-L810

Comment (graingert) [reply src/urllib3/connection.py:450]: fix is pretty much what cpython does: https://github.com/python/cpython/blob/dd69f734218ac5d3a551227069ac53ee09b0cd3e/Lib/http/client.py#L1444-L1455

Comment (graingert) [reply src/urllib3/connection.py:450]: this weird Exception/close responsibility is forced on urrlib3 by cpython http.client.HTTPConnection: https://github.com/python/cpython/blob/dd69f734218ac5d3a551227069ac53ee09b0cd3e/Lib/urllib/request.py#L1346-L1355

File: test/with_dummyserver/test_https.py [hunk 2]
```diff
@@ -329,6 +329,23 @@ def test_verified_with_bad_ca_certs(self) -> None:
...
+        with HTTPSConnectionPool(
+            self.host,
+            self.port,
+            cert_reqs="CERT_REQUIRED",
+            ca_certs=self.bad_ca_path,
+            ssl_minimum_version=self.tls_version(),
+        ) as https_pool:
+            conn = https_pool._get_conn()
+            try:
+                with pytest.raises(SSLError):
```
Comment (graingert) [inline test/with_dummyserver/test_https.py:342]:
```suggestion
                with pytest.raises(ssl.SSLError):
```

**Review by sethmlarson — APPROVED** [review]
LGTM, the test failures seem transient? Going to kick the CI to see if that's the case.

**Comment by pquentin** [comment]
Nice, the tests passed. Do we need a changelog entry, or are we going to pick it up from 1.26.x?

**Comment by sethmlarson** [comment]
@pquentin Yeah, we're backporting this one, no changelog needed.

**Comment by sethmlarson** [comment]
@graingert Thanks for this, would you like to create the backport PR too?

**Pull request merged**